src/compiler/utilities.ts

```diff
@@ -11591,3 +11593,35 @@ export function hasInferredType(node: Node): node is HasInferredType {
             return false;
     }
 }
+
+/** @internal */
+export function isEqualityOperatorKind(kind: SyntaxKind): kind is EqualityOperator {
+    switch (kind) {
+        case SyntaxKind.EqualsEqualsEqualsToken:
+        case SyntaxKind.EqualsEqualsToken:
+        case SyntaxKind.ExclamationEqualsEqualsToken:
+        case SyntaxKind.ExclamationEqualsToken:
+            return true;
+        default:
+            return false;
+    }
+}
+
+/** @internal */
+export function getContextualTypeFromParent(node: Expression, checker: TypeChecker, contextFlags?: ContextFlags): Type | undefined {
+    const parent = walkUpParenthesizedExpressions(node.parent);
+    switch (parent.kind) {
+        case SyntaxKind.NewExpression:
+            return checker.getContextualType(parent as NewExpression, contextFlags);
+        case SyntaxKind.BinaryExpression: {
+            const { left, operatorToken, right } = parent as BinaryExpression;
+            return isEqualityOperatorKind(operatorToken.kind)
+                ? checker.getTypeAtLocation(node === right ? left : right)
+                : checker.getContextualType(node, contextFlags);
+        }
+        case SyntaxKind.CaseClause:
+            return checker.getTypeAtLocation((parent as CaseClause).parent.parent.expression);

```

Comment:
It can't be moved to the checker. Those are specific to services and, for example, power renames in situations like:
```ts
type Foo = "a" | "b";

class C {
   p: Foo = "a";
   m() {
       if (this.p === "a") {}
   }
}
```

When u rename `"a"` to some other string that rename carries over to those other locations. But outside of situations like this the left/right operand isn't meant to contextually type the other one when equality comparator is involved.

Similarly, usually identifiers are not contextually-typed but this logic in services provides contextual typing for them.